File: Scala/src/main/scala/typeclasses/json/domain/Person.scala
package typeclasses.json.domain

import typeclasses.json.lib.JsonWriter.JsonWriterOps
import typeclasses.json.lib.{JsObject, JsonWriter}
import typeclasses.json.lib.JsonWriter.Instances._

final case class Person(name: String, married: Boolean, age: Int, email: String, children: Seq[String])

object Person {

  implicit val personWriter: JsonWriter[Person] = (p: Person) =>
    JsObject(Map(
      "name" -> p.name.toJson,
      "married" -> p.married.toJson,
      "age" -> p.age.toJson,
      "email" -> p.email.toJson,
      "children" -> p.children.toJson
    ))
}

File: Haskell/src/Person.hs
module Person where

import qualified Data.Map as M
import           Json

-- domain classes
--
data Person = Person
  { name     :: String
  , married  :: Bool
  , age      :: Int
  , email    :: String
  , children :: [String]
  }

-- type class instances for doamin classes
--
instance JsonWriter Person where
  toJson person = JsObject (M.fromList tuples)
    where
      tuples :: [(String, Json)]
      tuples =
        [ ("name", toJson $ name person)
        , ("married", toJson $ married person)
        , ("age", toJson $ age person)
        , ("email", toJson $ email person)
        , ("children", toJson $ children person)
        ]
